### ISSUE TITLE
[ORM Mode] Cannot populate by field name if alias exists in object

### ISSUE THREAD

**Issue by lukapeschke** [pydantic/pydantic#1463]
# Bug

**Description**:  When trying to populate by field name a field with an alias in ORM mode, validation will fail if the ORM model has a field with the name of the alias which is not of the expected type. The validation will fail even if the ORM field corresponding to the pydantic field's name is valid.

Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`:
```
         pydantic version: 1.5.1
        pydantic compiled: True
             install path: /home/lpeschke/work/data/fastapi-evaluation/venv/lib/python3.8/site-packages/pydantic
           python version: 3.8.2 (default, Feb 28 2020, 00:00:00)  [GCC 10.0.1 20200216 (Red Hat 10.0.1-0.8)]
                 platform: Linux-5.6.7-300.fc32.x86_64-x86_64-with-glibc2.2.5
 optional deps. installed: []
```

Example snippet:

```py
import typing

import pydantic
import sqlalchemy as sa
from sqlalchemy.ext.declarative import declarative_base


class MyModel(pydantic.BaseModel):
    metadata_: typing.Optional[typing.Dict[str, str]] = pydantic.Field(
        alias="metadata"
    )

    class Config:
        orm_mode = True
        allow_population_by_field_name = True


BaseModel = declarative_base()


class Model(BaseModel):

    __tablename__ = "my_table"
    id = sa.Column("id", sa.Integer, primary_key=True)
    # "metadata" is reserved by SQLAlchemy, hence the "_"
    metadata_ = sa.Column("metadata", sa.JSON, nullable=True, none_as_null=True)


sql_model = Model(metadata_={"key": "val"}, id=1)

pydantic_model = MyModel.from_orm(sql_model)
```

This snippet raises:
 ```
pydantic.error_wrappers.ValidationError: 1 validation error for MyModel
metadata
  value is not a valid dict (type=type_error.dict)
```
**Expected behavior**: `MyModel` should be populated with the value of `metadata_`.

I'd be glad to try to provide a fix for this, either by trying to validate the field name if the alias is invalid, or by providing a config option that specifies to ignore the alias and only use the field name for validation (`force_population_by_field_name` maybe ?)

Thanks for you work on Pydantic, It's been very pleasant to use so far!

**Comment by samuelcolvin** [comment]
Humm, I see the problem but I think the fix might be quite tricky since in the end `from_orm` uses the same logic as normal parsing which gives priority to aliases over field names.

Perhaps (and this might need to wait until v2, I'm not sure):

We add a `populate_by` argument to `parse_obj` and `from_orm` than can have values of `alias`, `field_name` or `either`. `alias` would be the same as the current default, `either` would be the same as with `allow_population_by_field_name = True` and `field_name` would be new behaviour which would solve your problem.

I guess to be consistent we should also add this to `Config` and depreciate `force_population_by_field_name`.

**Comment by lukapeschke** [comment]
Thanks for the reply! Your solution would be fine with me. Would you mind if I hack around on a PR ?

What are your plans for v2 ? Are there some major/breaking changes we're looking forward to, or will you consider a 1.x version that has evolved enough since v1.0 to be a v2 ?

**Comment by samuelcolvin** [comment]
No v2 will break some atuff, have a look at the milestone, but there are a few other changes I need to create issues for.

My all means have a hack, I'll happily review.

**Comment by lukapeschke** [comment]
Hi @samuelcolvin, I tinkered around with the code this week-end, and since aliases have priority over field names, my problem can be solved that way:

```py
import typing

import pydantic
import sqlalchemy as sa
from sqlalchemy.ext.declarative import declarative_base


class MyModel(pydantic.BaseModel):
    metadata: typing.Optional[typing.Dict[str, str]] = pydantic.Field(
        alias="metadata_"
    )

    class Config:
        orm_mode = True


BaseModel = declarative_base()


class Model(BaseModel):

    __tablename__ = "my_table"
    id = sa.Column("id", sa.Integer, primary_key=True)
    # "metadata" is reserved by SQLAlchemy, hence the "_"
    metadata_ = sa.Column("metadata", sa.JSON, nullable=True, none_as_null=True)


sql_model = Model(metadata_={"key": "val"}, id=1)

pydantic_model = MyModel.from_orm(sql_model)
print(pydantic_model.dict(by_alias=True))
``` 
I didn't realize that it was possible to specify which keys should be used when formatting the model.

So I don't know if the `populate_by` argument is required anymore since I believe every variant's behaviour can already be implemented ?

It might be worth to add an example showing this to Usage/Models/ORM Mode though, as there are high chances that people will create a column whose name is a reserved keyword at some point. Let me know if you think this makes sense, and I'll add it.

Anyway, thank you for your time on this :smile:

**Comment by samuelcolvin** [comment]
PR with better docs would be great.

Let's leave this open and keep thinking about `populate_by`.

**Comment by kevlarr** [comment]
A `populate_by="field_name"` option could actually be pretty handy.

We have a lot of SQLAlchemy models with `id` properties that correspond to database primary keys that we don't expose in JSON responses; instead, these models have other properties (usually some string id) that our front-end *thinks* is the primary id. Using `alias` doesn't quite get us there without manual intervention since it allows populating by **either** the alias **or** the field name.

For instance, given this:

```python
from pydantic import BaseModel as PydanticModel, Field
from myproject.basemodel import DatabaseModel

class Person(DatabaseModel):
    id = sa.Column(sa.Integer, primary_key=True)
    some_string_id = sa.Column(sa.Text, nullable=False)

class OrmBase(PydanticModel):
    class Config:
        orm_mode = True
        allow_population_by_field_name = True

class PersonSchema1(OrmBase):
    id: str = Field(..., alias="some_string_id")

class PersonSchema2(OrmBase):
    some_string_id: str = Field(..., alias="id")
```

Basically, given `Person(id=5, some_string_id="a-pretend-id")` I want to see `{"id": "a-pretend-id"}`...

Neither of these really get us there, because the first serializes as `{"some_string_id": "a-pretend-id"}` while the second serializes as `{"id": "5"}`, so I use a root validator with `pre=True` to get the right fields from the models/relationships as needed, but it's kind of verbose.

I think that being able to do...

```python
class PersonSchema1(OrmBase):
    some_string_id: str = Field(..., alias="id", populate_by="field_name")
```
... would indeed give us what we want in a very concise, very declarative manner.

**Comment by mike-hart** [comment]
Hi there. We're also using pydantic to parse objects into a well-defined structure, so we have a nested schema. We use aliases in order to permit loading from the odd names present in the source data. The objects are then serialised to a DB. The problem then comes when attempting to load them back (as above, and others on tickets I've seen in the past but don't have reference to right now).

The idea of a `populate_by=Populate.field_name` for the various `Model.from_*` helper methods would be great. It would maintain safety by not permitting _either_ alias-name _or_ field-name to be used whenever an instance is created (the right interface for the right occasion). We can't use the simple approach of subclassing since there are nested objects in play which means I'd need to subclass the entire tree, then re-link the references so that all the subclass fields refer to similarly subclassed definitions.

```python
class A(BaseModel):
    a: int = Field(alias="aval")

class B(BaseModel):
    x: int
    y: str = Field(alias="izzy")
    z: list[A] = Field(default_factory=list)

class APrime(A):
    class Config:
        allow_population_by_field_name = True

class BPrime(B):
    class Config:
        allow_population_by_field_name = True
    z: list[APrime] = Field(default_factory=list)
```
Sadly, even with the hacks above, mypy hates this because:
```
error: Incompatible types in assignment (expression has type "List[APrime]", base class "B" defined the type as "List[A]")
```
And both `APrime` and `BPrime` need the `Config` because otherwise when you try to restore something like:
```json
{
  "x": 1,
  "y": "hello",
  "z": [
    {"a": 2}
  ]
}
```
from a database (stored via `b_instance.json()`) when it tries to load an instance of `A` it relies on the alias.

Although not (specifically - but indirectly since a lot of tools rely on being given a python dict to serialise) related to this particular issue I also have desires for a jsonable version of `.dict()` (such that I could use things like `aiohttp.post(some_url, json=pydantic_obj.some_method())` or some database JSON `INSERT` and `UPDATE` serialisation handlers) rather than needing to get hacky (or dump/load before being able to proceed).

Right now we're using `allow_population_by_field_name` in a base-class, but it feels like the wrong tool for the intended outcome. Is the populate-by functionality definitely something coming in 2.x? If so, is there an expected timeline?

Thanks.

**Comment by guya-pecan** [comment]
+1 for this idea, and leaving this here for anyone else that might come across my problem which was quite similar:

in our project we had to store a relationship between a Resource and its Metadata and expose the SQL Alchemy ORM model via the API.

SQL Alchemy forbids using Metadata as a key, so we had to use an alias to map it properly to the Pydantic model.

However that broke our API schemas, so we had to find a solution -

If you only need to populate by the alias and only expose the Field name on the endpoint response, the `response_model_by_alias` flag is just for you.

```
@router.get(
    "/resources",
    response_model=MyPublicResourceWithAliasModel,
    response_model_by_alias=False,
)
```
